spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerPersistentEntityImpl.java

```diff
@@ -94,12 +103,49 @@ private Expression detectExpression() {
 	public void addPersistentProperty(SpannerPersistentProperty property) {
 		addPersistentPropertyToPersistentEntity(property);
 		this.columnNames.add(property.getColumnName());
+
+		if (property.getPrimaryKeyOrder() != null
+				&& property.getPrimaryKeyOrder().isPresent()) {
+			int order = property.getPrimaryKeyOrder().getAsInt();
+			if (this.primaryKeyParts.containsKey(order)) {
+				throw new SpannerDataException(
+						"Two properties were annotated with the same primary key order: "
+								+ property.getColumnName() + " and "
+								+ this.primaryKeyParts.get(order).getColumnName());
+			}
+			this.primaryKeyParts.put(order, property);
+		}
 	}
 
 	private void addPersistentPropertyToPersistentEntity(SpannerPersistentProperty property) {
 		super.addPersistentProperty(property);
 	}
 
+	@Override
+	public SpannerPersistentProperty getIdProperty() {
+		return new SpannerCompositeKeyProperty(this, getPrimaryKeyProperties());
```

Comment:
Staying in the same property implementation realm allows removing specializations such as this one.